File: android/app/src/main/java/com/zillit/drive/data/remote/dto/DriveResponses.kt
package com.zillit.drive.data.remote.dto

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

// ─── Generic API Response Wrapper ───

@JsonClass(generateAdapter = true)
data class ApiResponse<T>(
    @Json(name = "status") val status: Int,
    @Json(name = "message") val message: String? = null,
    @Json(name = "messageElements") val messageElements: List<String>? = null,
    @Json(name = "data") val data: T? = null
) {
    val isSuccess: Boolean get() = status == 1
}

// ─── Permissions DTO ───

@JsonClass(generateAdapter = true)
data class UserPermissionsDto(
    @Json(name = "can_view") val canView: Boolean? = null,
    @Json(name = "can_edit") val canEdit: Boolean? = null,
    @Json(name = "can_download") val canDownload: Boolean? = null,
    @Json(name = "can_delete") val canDelete: Boolean? = null
)

// ─── File DTOs ───

@JsonClass(generateAdapter = true)
data class DriveFileDto(
    @Json(name = "_id") val id: String,
    @Json(name = "file_name") val fileName: String,
    @Json(name = "file_extension") val fileExtension: String? = null,
    @Json(name = "file_size_bytes") val fileSizeBytes: Long = 0,
    @Json(name = "mime_type") val mimeType: String? = null,
    @Json(name = "folder_id") val folderId: String? = null,
    @Json(name = "file_path") val filePath: String? = null,
    @Json(name = "description") val description: String? = null,
    @Json(name = "created_by") val createdBy: String? = null,
    @Json(name = "created_on") val createdOn: Long = 0,
    @Json(name = "updated_on") val updatedOn: Long = 0,
    @Json(name = "deleted_on") val deletedOn: Long = 0,
    @Json(name = "attachments") val attachments: List<AttachmentDto>? = null,
    @Json(name = "_userPermissions") val userPermissions: UserPermissionsDto? = null
)

@JsonClass(generateAdapter = true)
data class AttachmentDto(
    @Json(name = "media") val media: String? = null,
    @Json(name = "thumbnail") val thumbnail: String? = null
)

// ─── Folder DTOs ───

@JsonClass(generateAdapter = true)
data class DriveFolderDto(
    @Json(name = "_id") val id: String,
    @Json(name = "folder_name") val folderName: String,
    @Json(name = "parent_folder_id") val parentFolderId: String? = null,
    @Json(name = "description") val description: String? = null,
    @Json(name = "created_by") val createdBy: String? = null,
    @Json(name = "created_on") val createdOn: Long = 0,
    @Json(name = "updated_on") val updatedOn: Long = 0,
    @Json(name = "deleted_on") val deletedOn: Long = 0,
    @Json(name = "file_count") val fileCount: Int = 0,
    @Json(name = "folder_count") val folderCount: Int = 0,
    @Json(name = "_userPermissions") val userPermissions: UserPermissionsDto? = null
)

// ─── Contents DTO ───

@JsonClass(generateAdapter = true)
data class DriveContentsDto(
    @Json(name = "items") val items: List<DriveContentItemDto> = emptyList(),
    @Json(name = "pagination") val pagination: PaginationDto? = null,
    @Json(name = "counts") val counts: CountsDto? = null
)

@JsonClass(generateAdapter = true)
data class DriveContentItemDto(
    @Json(name = "_id") val id: String,
    @Json(name = "type") val type: String? = null,
    @Json(name = "is_folder") val isFolder: Boolean? = null,
    @Json(name = "name") val name: String? = null,
    @Json(name = "folder_name") val folderName: String? = null,
    @Json(name = "file_name") val fileName: String? = null,
    @Json(name = "file_extension") val fileExtension: String? = null,
    @Json(name = "file_size_bytes") val fileSizeBytes: Long = 0,
    @Json(name = "mime_type") val mimeType: String? = null,
    @Json(name = "folder_id") val folderId: String? = null,
    @Json(name = "parent_folder_id") val parentFolderId: String? = null,
    @Json(name = "file_path") val filePath: String? = null,
    @Json(name = "description") val description: String? = null,
    @Json(name = "created_by") val createdBy: String? = null,
    @Json(name = "created_on") val createdOn: Long = 0,
    @Json(name = "updated_on") val updatedOn: Long = 0,
    @Json(name = "deleted_on") val deletedOn: Long = 0,
    @Json(name = "file_count") val fileCount: Int = 0,
    @Json(name = "folder_count") val folderCount: Int = 0,
    @Json(name = "attachments") val attachments: List<AttachmentDto>? = null,
    @Json(name = "_userPermissions") val userPermissions: UserPermissionsDto? = null
) {
    val resolvedIsFolder: Boolean
        get() = isFolder == true || type == "folder"

    val resolvedName: String
        get() = name ?: folderName ?: fileName ?: ""
}

@JsonClass(generateAdapter = true)
data class PaginationDto(
    @Json(name = "total") val total: Int = 0,
    @Json(name = "limit") val limit: Int = 50,
    @Json(name = "offset") val offset: Int = 0,
    @Json(name = "has_more") val hasMore: Boolean = false
)

@JsonClass(generateAdapter = true)
data class CountsDto(
    @Json(name = "folders") val folders: Int = 0,
    @Json(name = "files") val files: Int = 0,
    @Json(name = "total") val total: Int = 0
)

// ─── Upload DTOs ───

@JsonClass(generateAdapter = true)
data class InitiateUploadRequest(
    @Json(name = "file_name") val fileName: String,
    @Json(name = "file_size_bytes") val fileSizeBytes: Long,
    @Json(name = "folder_id") val folderId: String? = null,
    @Json(name = "mime_type") val mimeType: String,
    @Json(name = "description") val description: String? = null
)

@JsonClass(generateAdapter = true)
data class UploadSessionDto(
    @Json(name = "upload_id") val uploadId: String,
    @Json(name = "s3_upload_id") val s3UploadId: String? = null,
    @Json(name = "file_name") val fileName: String? = null,
    @Json(name = "file_size_bytes") val fileSizeBytes: Long = 0,
    @Json(name = "chunk_size") val chunkSize: Long = 0,
    @Json(name = "total_parts") val totalParts: Int = 0,
    @Json(name = "presigned_urls") val presignedUrls: List<PresignedUrlDto> = emptyList()
)

@JsonClass(generateAdapter = true)
data class PresignedUrlDto(
    @Json(name = "part_number") val partNumber: Int,
    @Json(name = "url") val url: String
)

@JsonClass(generateAdapter = true)
data class CompleteUploadRequest(
    @Json(name = "parts") val parts: List<UploadPartDto>
)

@JsonClass(generateAdapter = true)
data class UploadPartDto(
    @Json(name = "part_number") val partNumber: Int,
    @Json(name = "etag") val etag: String
)

// ─── Access DTOs ───

@JsonClass(generateAdapter = true)
data class FolderAccessDto(
    @Json(name = "user_id") val userId: String,
    @Json(name = "folder_id") val folderId: String? = null,
    @Json(name = "role") val role: String,
    @Json(name = "inherited") val inherited: Boolean = false
)

@JsonClass(generateAdapter = true)
data class FileAccessDto(
    @Json(name = "user_id") val userId: String,
    @Json(name = "file_id") val fileId: String? = null,
    @Json(name = "can_view") val canView: Boolean = true,
    @Json(name = "can_edit") val canEdit: Boolean = false,
    @Json(name = "can_download") val canDownload: Boolean = false
)

@JsonClass(generateAdapter = true)
data class UpdateAccessRequest(
    @Json(name = "entries") val entries: List<FolderAccessDto>,
    @Json(name = "replace_existing") val replaceExisting: Boolean = true
)

@JsonClass(generateAdapter = true)
data class UpdateFileAccessRequest(
    @Json(name = "entries") val entries: List<FileAccessDto>
)

// ─── Tag DTOs ───

@JsonClass(generateAdapter = true)
data class DriveTagDto(
    @Json(name = "_id") val id: String,
    @Json(name = "name") val name: String,
    @Json(name = "color") val color: String? = null,
    @Json(name = "project_id") val projectId: String? = null
)

@JsonClass(generateAdapter = true)
data class TagAssignRequest(
    @Json(name = "tag_id") val tagId: String,
    @Json(name = "item_id") val itemId: String,
    @Json(name = "item_type") val itemType: String // "file" or "folder"
)

// ─── Comment DTOs ───

@JsonClass(generateAdapter = true)
data class DriveCommentDto(
    @Json(name = "_id") val id: String,
    @Json(name = "file_id") val fileId: String? = null,
    @Json(name = "user_id") val userId: String? = null,
    @Json(name = "text") val text: String,
    @Json(name = "created_on") val createdOn: Long = 0,
    @Json(name = "updated_on") val updatedOn: Long = 0
)

// ─── Version DTOs ───

@JsonClass(generateAdapter = true)
data class DriveVersionDto(
    @Json(name = "_id") val id: String,
    @Json(name = "file_id") val fileId: String? = null,
    @Json(name = "version_number") val versionNumber: Int = 0,
    @Json(name = "file_size_bytes") val fileSizeBytes: Long = 0,
    @Json(name = "file_path") val filePath: String? = null,
    @Json(name = "created_by") val createdBy: String? = null,
    @Json(name = "created_on") val createdOn: Long = 0
)

// ─── Activity DTOs ───

@JsonClass(generateAdapter = true)
data class DriveActivityDto(
    @Json(name = "_id") val id: String,
    @Json(name = "action") val action: String,
    @Json(name = "item_id") val itemId: String? = null,
    @Json(name = "item_type") val itemType: String? = null,
    @Json(name = "user_id") val userId: String? = null,
    @Json(name = "details") val details: String? = null,
    @Json(name = "created_on") val createdOn: Long = 0
)

// ─── Storage DTO ───

@JsonClass(generateAdapter = true)
data class StorageUsageDto(
    @Json(name = "used_bytes") val usedBytes: Long = 0,
    @Json(name = "total_bytes") val totalBytes: Long = 0,
    @Json(name = "file_count") val fileCount: Int = 0
)

// ─── Share Link DTO ───

@JsonClass(generateAdapter = true)
data class ShareLinkDto(
    @Json(name = "url") val url: String,
    @Json(name = "expires_at") val expiresAt: String? = null
)

// ─── Stream URL DTO ───

@JsonClass(generateAdapter = true)
data class StreamUrlDto(
    @Json(name = "url") val url: String
)

// ─── Favorites ───

@JsonClass(generateAdapter = true)
data class ToggleFavoriteRequest(
    @Json(name = "item_id") val itemId: String,
    @Json(name = "item_type") val itemType: String
)

@JsonClass(generateAdapter = true)
data class FavoriteIdsDto(
    @Json(name = "file_ids") val fileIds: List<String> = emptyList(),
    @Json(name = "folder_ids") val folderIds: List<String> = emptyList()
)

// ─── Bulk Operations ───

@JsonClass(generateAdapter = true)
data class BulkDeleteRequest(
    @Json(name = "items") val items: List<BulkItemDto>
)

@JsonClass(generateAdapter = true)
data class BulkMoveRequest(
    @Json(name = "items") val items: List<BulkItemDto>,
    @Json(name = "target_folder_id") val targetFolderId: String?
)

@JsonClass(generateAdapter = true)
data class BulkItemDto(
    @Json(name = "id") val id: String,
    @Json(name = "type") val type: String // "file" or "folder"
)

@JsonClass(generateAdapter = true)
data class BulkDownloadRequest(
    @Json(name = "file_ids") val fileIds: List<String>
)

// ─── Editor ───

@JsonClass(generateAdapter = true)
data class EditorPageTokenDto(
    @Json(name = "token") val token: String
)

// ─── Trash ───

@JsonClass(generateAdapter = true)
data class TrashItemDto(
    @Json(name = "_id") val id: String,
    @Json(name = "type") val type: String, // "file" or "folder"
    @Json(name = "name") val name: String,
    @Json(name = "deleted_on") val deletedOn: Long = 0,
    @Json(name = "file") val file: DriveFileDto? = null,
    @Json(name = "folder") val folder: DriveFolderDto? = null
)

File: ios/ZillitDrive/Data/Remote/DTOs/DriveResponses.swift
import Foundation

// MARK: - Generic API Response

struct APIResponse<T: Decodable>: Decodable {
    let status: Int
    let message: String?
    let messageElements: [String]?
    let data: T?

    var isSuccess: Bool { status == 1 }
}

// MARK: - Permissions DTO

struct UserPermissionsDTO: Codable {
    let canView: Bool?
    let canEdit: Bool?
    let canDownload: Bool?
    let canDelete: Bool?

    enum CodingKeys: String, CodingKey {
        case canView = "can_view"
        case canEdit = "can_edit"
        case canDownload = "can_download"
        case canDelete = "can_delete"
    }
}

// MARK: - File DTOs

struct DriveFileDTO: Codable, Identifiable {
    let id: String
    let fileName: String
    let fileExtension: String?
    let fileSizeBytes: Int64?
    let mimeType: String?
    let folderId: String?
    let filePath: String?
    let description: String?
    let createdBy: String?
    let createdOn: Int64?
    let updatedOn: Int64?
    let deletedOn: Int64?
    let attachments: [AttachmentDTO]?
    var userPermissions: UserPermissionsDTO? = nil

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case fileName = "file_name"
        case fileExtension = "file_extension"
        case fileSizeBytes = "file_size_bytes"
        case mimeType = "mime_type"
        case folderId = "folder_id"
        case filePath = "file_path"
        case description
        case createdBy = "created_by"
        case createdOn = "created_on"
        case updatedOn = "updated_on"
        case deletedOn = "deleted_on"
        case attachments
        case userPermissions = "_userPermissions"
    }
}

struct AttachmentDTO: Codable {
    let media: String?
    let thumbnail: String?
}

// MARK: - Folder DTOs

struct DriveFolderDTO: Codable, Identifiable {
    let id: String
    let folderName: String
    let parentFolderId: String?
    let description: String?
    let createdBy: String?
    let createdOn: Int64?
    let updatedOn: Int64?
    let deletedOn: Int64?
    let fileCount: Int?
    let folderCount: Int?
    var userPermissions: UserPermissionsDTO? = nil

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case folderName = "folder_name"
        case parentFolderId = "parent_folder_id"
        case description
        case createdBy = "created_by"
        case createdOn = "created_on"
        case updatedOn = "updated_on"
        case deletedOn = "deleted_on"
        case fileCount = "file_count"
        case folderCount = "folder_count"
        case userPermissions = "_userPermissions"
    }
}

// MARK: - Contents DTO (from /v2/drive/folders/contents)

struct DriveContentsDTO: Codable {
    let items: [DriveContentItemDTO]?
    let pagination: PaginationDTO?
    let counts: CountsDTO?
}

struct DriveContentItemDTO: Codable, Identifiable {
    let id: String
    let type: String?           // "folder" or "file"
    let isFolder: Bool?
    let name: String?
    let dateModified: Int64?
    let size: Int64?

    // Folder fields
    let folderName: String?
    let folderPath: String?
    let parentFolderId: String?

    // File fields
    let fileName: String?
    let fileExtension: String?
    let fileSizeBytes: Int64?
    let mimeType: String?
    let folderId: String?
    let filePath: String?
    let attachments: [AttachmentDTO]?

    // Common fields
    let description: String?
    let createdBy: String?
    let createdOn: Int64?
    let updatedOn: Int64?
    let deletedOn: Int64?
    let fileCount: Int?
    let folderCount: Int?
    var userPermissions: UserPermissionsDTO? = nil

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case type
        case isFolder = "is_folder"
        case name
        case dateModified = "date_modified"
        case size
        case folderName = "folder_name"
        case folderPath = "folder_path"
        case parentFolderId = "parent_folder_id"
        case fileName = "file_name"
        case fileExtension = "file_extension"
        case fileSizeBytes = "file_size_bytes"
        case mimeType = "mime_type"
        case folderId = "folder_id"
        case filePath = "file_path"
        case attachments
        case description
        case createdBy = "created_by"
        case createdOn = "created_on"
        case updatedOn = "updated_on"
        case deletedOn = "deleted_on"
        case fileCount = "file_count"
        case folderCount = "folder_count"
        case userPermissions = "_userPermissions"
    }
}

struct PaginationDTO: Codable {
    let total: Int?
    let limit: Int?
    let offset: Int?
    let hasMore: Bool?

    enum CodingKeys: String, CodingKey {
        case total, limit, offset
        case hasMore = "has_more"
    }
}

struct CountsDTO: Codable {
    let folders: Int?
    let files: Int?
    let total: Int?
}

// MARK: - Upload DTOs

struct UploadSessionDTO: Codable {
    let uploadId: String
    let s3UploadId: String?
    let fileName: String?
    let fileSizeBytes: Int64?
    let chunkSize: Int64?
    let totalParts: Int?
    let presignedUrls: [PresignedUrlDTO]?

    enum CodingKeys: String, CodingKey {
        case uploadId = "upload_id"
        case s3UploadId = "s3_upload_id"
        case fileName = "file_name"
        case fileSizeBytes = "file_size_bytes"
        case chunkSize = "chunk_size"
        case totalParts = "total_parts"
        case presignedUrls = "presigned_urls"
    }
}

struct PresignedUrlDTO: Codable {
    let partNumber: Int
    let url: String

    enum CodingKeys: String, CodingKey {
        case partNumber = "part_number"
        case url
    }
}

// MARK: - Access DTOs

struct FolderAccessDTO: Codable {
    let userId: String
    let folderId: String?
    let role: String
    let inherited: Bool?

    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case folderId = "folder_id"
        case role, inherited
    }
}

struct FileAccessDTO: Codable {
    let userId: String
    let fileId: String?
    let canView: Bool?
    let canEdit: Bool?
    let canDownload: Bool?

    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case fileId = "file_id"
        case canView = "can_view"
        case canEdit = "can_edit"
        case canDownload = "can_download"
    }
}

// MARK: - Tag DTOs

struct DriveTagDTO: Codable, Identifiable {
    let id: String
    let name: String
    let color: String?
    let projectId: String?

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case name, color
        case projectId = "project_id"
    }
}

// MARK: - Comment DTOs

struct DriveCommentDTO: Codable, Identifiable {
    let id: String
    let fileId: String?
    let userId: String?
    let text: String
    let createdOn: Int64?
    let updatedOn: Int64?

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case fileId = "file_id"
        case userId = "user_id"
        case text
        case createdOn = "created_on"
        case updatedOn = "updated_on"
    }
}

// MARK: - Version DTOs

struct DriveVersionDTO: Codable, Identifiable {
    let id: String
    let fileId: String?
    let versionNumber: Int?
    let fileSizeBytes: Int64?
    let filePath: String?
    let createdBy: String?
    let createdOn: Int64?

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case fileId = "file_id"
        case versionNumber = "version_number"
        case fileSizeBytes = "file_size_bytes"
        case filePath = "file_path"
        case createdBy = "created_by"
        case createdOn = "created_on"
    }
}

// MARK: - Activity DTOs

struct DriveActivityDTO: Codable, Identifiable {
    let id: String
    let action: String
    let itemId: String?
    let itemType: String?
    let userId: String?
    let details: String?
    let createdOn: Int64?

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case action
        case itemId = "item_id"
        case itemType = "item_type"
        case userId = "user_id"
        case details
        case createdOn = "created_on"
    }
}

// MARK: - Storage DTO

struct StorageUsageDTO: Codable {
    let usedBytes: Int64?
    let totalBytes: Int64?
    let fileCount: Int?

    enum CodingKeys: String, CodingKey {
        case usedBytes = "used_bytes"
        case totalBytes = "total_bytes"
        case fileCount = "file_count"
    }
}

// MARK: - Stream URL DTO

struct StreamUrlDTO: Codable {
    let url: String
}

// MARK: - Share Link DTO

struct ShareLinkDTO: Codable {
    let url: String
    let expiresAt: String?

    enum CodingKeys: String, CodingKey {
        case url
        case expiresAt = "expires_at"
    }
}

// MARK: - Favorite IDs

struct FavoriteIdsDTO: Codable {
    let fileIds: [String]?
    let folderIds: [String]?

    enum CodingKeys: String, CodingKey {
        case fileIds = "file_ids"
        case folderIds = "folder_ids"
    }
}
